src/libopensc/card-epass2003.c

```diff
@@ -1330,6 +1330,100 @@ epass2003_sm_wrap_apdu(struct sc_card *card, struct sc_apdu *plain, struct sc_ap
 	return SC_SUCCESS;
 }
 
+static int
+epass2003_check_response_mac_and_sw(struct sc_card *card, struct sc_apdu *sm)
+{
+	unsigned char iv[16];
+	unsigned char *data = NULL, *mac = NULL;
+	size_t blocksize, mac_len;
+	int ret = -1;
+	size_t taglen;
+	const u8 *tag;
+	epass2003_exdata *exdata;
+	unsigned char *in = sm->resp;
+	unsigned char *alt_in;
+	size_t inlen = sm->resplen;
+
+	/* card/ctx/drv_data is already checked by caller */
+	exdata = (epass2003_exdata *)card->drv_data;
+
+	/* compare BER-TLV encoded SW (TAG 0x99) and raw SW */
+	alt_in = in;
+	tag = sc_asn1_find_tag(card->ctx, alt_in, inlen, 0x99, &taglen);
+	if (tag == NULL || taglen != 2) {
+		/*
+		 * It seems that the EPASS2003 firmware has some problem with BER-TLV encoding.
+		 * There seems to be some proprietary fix for the faulty encoding in the
+		 * decrypt_response() function.  Instead of (correct) TLV 87 81 81 [01 ...]
+		 * incorrect TLV 87 81 [01 ...  ] is returned.
+		 */
+		if (0x01 == in[2] && 0x82 != in[1]) {
+			sc_log(card->ctx, "Workaround, wrong BER-TLV ?");
+			alt_in += in[1] + 2;
+			tag = sc_asn1_find_tag(card->ctx, alt_in, inlen, 0x99, &taglen);
+			if (tag == NULL || taglen != 2)
+				return ret;
+		} else
+			return ret;
+	}
+	if (sm->sw1 != tag[0] || sm->sw2 != tag[1])
+		return ret;
+
+	if (exdata->bFipsCertification) {
+		sc_log(card->ctx, "Warning, MAC is not checked");
```

Comment:
So FIPS forbids authenticated responses for cards!?

Reply:
Currently, the SM message is not checked at all - the driver also accepts the SM message with an invalid MAC. After applying this PR, the check will be performed for the MAC generated using DES_CBC_MAC or AES_CBC_MAC. If the token uses AES_CMAC (FIPS certified mode), any SM is accepted as before (without MAC check).   I will write the code that uses AES CMAC later (as a separate PR), it is necessary to solve several workarounds where apparently IV is not used in the MAC generation for some APDUs.

Reply:
OK, then lets postpone this for later.